lib/modem_key_mgmt/modem_key_mgmt.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <at_cmd.h>
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_params.h>
+#include <bsd_limits.h>
+#include <modem_key_mgmt.h>
+#include <stdio.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define MODEM_KEY_MGMT_OP_LS "AT%CMNG=1"
+#define MODEM_KEY_MGMT_OP_RD "AT%CMNG=2"
+#define MODEM_KEY_MGMT_OP_WR "AT%CMNG=0"
+#define MODEM_KEY_MGMT_OP_RM "AT%CMNG=3"
+
+#define AT_CMNG_PARAMS_COUNT 5
+#define AT_CMNG_CONTENT_INDEX 4
+
+static char scratch_buf[4096];
+static char at_cmee_strings[2][10] = {"AT+CMEE=0", "AT+CMEE=1"};
+
+static int cmee_active(void)
+{
+	int result;
+	char response[sizeof("+CMEE: X\r\n") + 1];
+
+	memset(response, 0, sizeof(response));
+
+	result = at_cmd_write("AT+CMEE?", response, sizeof(response), NULL);
+	if (result == 0) {
+		if (strchr(&response[6], '1')) {
+			return 1;
+		} else {
+			return 0;
+		}
+	}
+
+	return -EIO;
+}
+
+static int cmee_set(bool enable)
+{
+	return at_cmd_write(at_cmee_strings[(int)enable], NULL, 0, NULL);
+}
+
+static int translate_error(int err)
+{
+	switch (err) {
+	case 513:
+		return -ENOENT;
+	case 514:
+		return -EPERM;
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
```

Comment:
Instead of returning `-EFAULT` and calling this function only to translate non-listed `errno` values, we could `return err` here. That way we can unconditionally translate the return value of `write_at_cmd_with_cme_enabled` and simply return that, instead of translating only some errors, and returning directly others.

Reply:
> Instead of returning -EFAULT and calling this function only to translate non-listed errno values, we could return err here.
@lemrey , i'm not sure if i get it. What is non-listed `errno` values?

Reply:
`513`, `514`, `515`, `518` are unlisted `errno` values; from what I can see you expect these values to be returned from `at_cmd_write()` inside `write_at_cmd_with_cme_enabled`, .. do I get it right?

Reply:
Yes, i expect these from the `at_cmd_write()`. As the documentation says, the numbers mentioned are not `errno`s but CME errors, in the case where `state == AT_CMD_ERROR_CME`. Hence, why i would like to translate them to `errno` in this library, so that the user of this API does not have to have any relation to the specific AT commands used.

Reply:
@lemrey, I have tried to centralize the error translation to this one function now.